usr/src/uts/common/fs/zfs/vdev_raidz.c

```diff
@@ -1685,6 +1707,32 @@ vdev_raidz_asize(vdev_t *vd, uint64_t psize)
 	return (asize);
 }
 
+/*
+ * Converts an allocated size on a raidz vdev back to a logical block
+ * size. This is used in trimming to figure out the appropriate logical
+ * size to pass to vdev_raidz_map_alloc when splitting up extents of free
+ * space obtained from metaslabs. However, a range of free space on a
+ * raidz vdev might have originally consisted of multiple blocks and
+ * those, taken together with their skip blocks, might not always align
+ * neatly to a new vdev_raidz_map_alloc covering the entire unified
+ * range. So to ensure that the newly allocated raidz map *always* fits
+ * within the asize passed to this function and never exceeds it (since
+ * that might trim allocated data past it), we round it down to the
+ * nearest suitable multiple of the vdev ashift (hence the "_floor" in
+ * this function's name).
+ */
+static uint64_t
+vdev_raidz_psize_floor(vdev_t *vd, uint64_t asize)
+{
+	uint64_t psize = (asize / vd->vdev_children) *
+	    (vd->vdev_children - vd->vdev_nparity);
+
+	psize = P2ALIGN(psize, 1 << vd->vdev_top->vdev_ashift);
+	ASSERT(psize != 0);
+
+	return (psize);
+}
+
```

Comment:
This function needed re-working in order to work properly with raidz2 and raidz3.  Details in zfsonlinux/zfs#3656 (fix from @ironmann) [EDIT] (I'll try to dig up a proper link to the issue comment).  See dweeezil/zfs@7c8ba1b for details.